### ISSUE TITLE
Permission check spam.

### ISSUE THREAD

**Issue by mibby** [Esophose/PlayerParticles#9]
PlayerParticle v5.1
LuckPerms dev 670 https://ci.lucko.me/job/LuckPerms/changes

I enabled verbose logging for my permissions plugin to debug something and noticed that PlayerParticle spams permission checks repeatedly. Is there any way to optimize this to not check as frequently?

```
[06:44:52] [luckperms-0/INFO]: [LP] VB > mibby - playerparticles.* - undefined
[06:44:52] [luckperms-0/INFO]: [LP] VB > mibby - playerparticles.effect.* - undefined
[06:44:52] [luckperms-0/INFO]: [LP] VB > mibby - playerparticles.effect.none - true
```
https://paste.ubuntu.com/p/RcQGzwfCwZ/

**Comment by Esophose** [comment]
This is true that it checks permissions repeatedly, 20 times per second, per player who has an effect/style active. It does this to make sure they still have permission before displaying the particles. I suppose I could make it check less frequently, say, every 5 seconds instead, but I would still have to do these semi-freauent checks.

My thoughts were that permission checks were pretty quick and shouldn't be slowing down the program at all. I never actually checked this to make sure. If you can think of any way to improve the efficiency of it besides what I mentioned already, let me know.

**Comment by mibby** [comment]
> per player who has an effect/style active

I didn't have any effects or styles active when I was debugging myself. :( But if you can reduce the number of times it checks and perhaps the frequency too, as well as if possible only when you have a particle / effect selected, I'd greatly appreciate it!

**Comment by mibby** [comment]
Any update to this issue? :(

**Comment by Esophose** [comment]
I don't have any updates for you. I only work on the plugin in my free time and I've been pretty busy lately. This doesn't seem like a very large issue to fix immediately since it doesn't cause any significant performance issues.

**Comment by mibby** [comment]
Okay, no worries. I just hope it can be fixed sometime in the near future, as it does make debugging permissions tedious in-game and from a verbose log output.

While it probably doesn't have a performance impact with a single person, surely it could cause a noticeable hit with hundreds of players online all being checked hundreds of times a second? It's the only plugin I have that constantly spams permission checks. :(

**Comment by Esophose** [comment]
The only way you can reduce the number of permission checks would be to not use the plugin for now. In v5.2 I will only apply permission checks when the user executes a command or joins the server. That should drastically reduce the number of checks.

**Comment by mibby** [comment]
Wonderful. I look forward to v5.2's release!

**Comment by Esophose** [comment]
What are your thoughts on me only updating the plugin for 1.13 from here on out? The 1.13 API brought a lot of changes that will be a pain to deal with having multiple versions of the plugin.

**Comment by mibby** [comment]
Personally I am perfectly fine with that as I always tend to stay on the latest supported MC and Spigot version. However I'm holding onto 1.12.2 until 1.13.1 releases due to some crippling bugs that need to be fixed first (in spigot and MC client) and some outdated plugins that are pending replacement / update rewrites.

If you have some form of analytics like mcstats, you could check what version people are using PlayerParticles on. But I don't have a problem with you only supporting 1.13+ hereon-out due to the particle name changes and removals.

If the permission check update isn't too extensive to backport, you could push that as a final build for older MC versions just so you have something stable that doesn't cause any performance impact with lots of players or bothersome permission debugging.

**Comment by Esophose** [comment]
Alright, thanks for the input. I'll include support for 1.9+ as long as the API changes don't get too crazy since I do realize Spigot is going to be in the extreme caution phase for a while.

**Comment by Esophose** [comment]
Issue resolved, permissions will now only be checked on command execution or tab completion. If the amount of permission checks are still a problem, please let me know. This change will be released with PlayerParticles v5.2 within the next two days.
